### PR TITLE
feat: expose query plans through response extensions field

### PR DESCRIPTION
**Pull request by jensneuse**

### CODE REVIEW

**Review by devsergiy — APPROVED** [review]

File: router/core/factoryresolver.go [hunk 1]
```diff
@@ -447,6 +450,18 @@ func (l *Loader) Load(engineConfig *nodev1.EngineConfiguration, subgraphs []*nod
...
+func (l *Loader) subgraphName(subgraphs []*nodev1.Subgraph, dataSourceID string) string {
+	i := slices.IndexFunc(subgraphs, func(s *nodev1.Subgraph) bool {
+		return s.Id == dataSourceID
+	})
+
+	if i != -1 {
+		return subgraphs[i].Name
+	}
+
+	return dataSourceID
```
Comment (StarpTech) [inline router/core/factoryresolver.go:462]: This should never happen. We should return empty string here to indicate that something is wrong.

File: router/core/graphql_prehandler.go [hunk 2]
```diff
@@ -646,3 +632,62 @@ func (h *PreHandler) flushMetrics(ctx context.Context, requestLogger *zap.Logger
 
 	requestLogger.Debug("Metrics flushed", zap.Duration("duration", time.Since(now)))
 }
+
+func (h *PreHandler) parseRequestOptions(r *http.Request, clientInfo *ClientInfo, requestLogger *zap.Logger) (resolve.ExecutionOptions, resolve.TraceOptions, error) {
```
Comment (StarpTech) [inline router/core/graphql_prehandler.go:636]: Good refactoring.

File: router/core/operation_planner.go [hunk 3]
```diff
@@ -85,32 +91,42 @@ func (p *OperationPlanner) preparePlan(requestOperationName []byte, requestOpera
...
 
-	if traceOptions.Enable {
-		// if we have tracing enabled we always prepare a new plan
-		// this is because we're writing trace data to the plan
-		requestOperationNameBytes := unsafebytes.StringToBytes(opContext.Name())
-		prepared, err := p.preparePlan(requestOperationNameBytes, opContext.Content())
+	// if we have tracing enabled or want to include a query plan in the response we always prepare a new plan
+	// this is because in case of tracing, we're writing trace data to the plan
+	// in case of including the query plan, we don't want to cache this additional overhead
+	skipCache := options.TraceOptions.Enable || options.ExecutionOptions.IncludeQueryPlanInResponse
```
Comment (StarpTech) [inline router/core/operation_planner.go:126]: Do we respect this condition in the OTEL attribute?

File: router/pkg/config/config.go [hunk 4]
```diff
@@ -218,7 +220,7 @@ type EngineExecutionConfiguration struct {
 	EpollKqueueConnBufferSize              int                      `envDefault:"128" env:"ENGINE_EPOLL_KQUEUE_CONN_BUFFER_SIZE" yaml:"epoll_kqueue_conn_buffer_size,omitempty"`
 	WebSocketReadTimeout                   time.Duration            `envDefault:"5s" env:"ENGINE_WEBSOCKET_READ_TIMEOUT" yaml:"websocket_read_timeout,omitempty"`
 	ExecutionPlanCacheSize                 int64                    `envDefault:"1024" env:"ENGINE_EXECUTION_PLAN_CACHE_SIZE" yaml:"execution_plan_cache_size,omitempty"`
-	MinifySubgraphOperations               bool                     `envDefault:"false" env:"ENGINE_MINIFY_SUBGRAPH_OPERATIONS" yaml:"minify_subgraph_operations"`
+	MinifySubgraphOperations               bool                     `envDefault:"true" env:"ENGINE_MINIFY_SUBGRAPH_OPERATIONS" yaml:"minify_subgraph_operations"`
```
Comment (StarpTech) [inline router/pkg/config/config.go:223]: For the future, I'd propose to make those change in different PR's. This makes it easier to track down issues.

File: router/pkg/config/config.schema.json [hunk 5]
```diff
@@ -1381,6 +1386,16 @@
             "enable_normalization_cache_response_header": {
               "type": "boolean",
               "description": "Enable the normalization cache response header. The normalization cache response header is used to cache the normalized operations in the client."
+            },
+            "always_include_query_plan": {
+              "type": "boolean",
+              "default": false,
+              "description": "Always include the query plan in the response."
+            },
+            "always_skip_loader": {
```
Comment (StarpTech) [inline router/pkg/config/config.schema.json:1395]: I'd like to go with `skip_loader` "always" is implicit when you set it.

File: router/pkg/config/config.schema.json [hunk 6]
```diff
@@ -1381,6 +1386,16 @@
             "enable_normalization_cache_response_header": {
               "type": "boolean",
               "description": "Enable the normalization cache response header. The normalization cache response header is used to cache the normalized operations in the client."
+            },
+            "always_include_query_plan": {
```
Comment (StarpTech) [inline router/pkg/config/config.schema.json:1390]: Same here.

File: router/pkg/otel/attributes.go [hunk 7]
```diff
@@ -32,6 +32,7 @@ const (
 	WgAcquireResolverWaitTimeMs        = attribute.Key("wg.engine.resolver.wait_time_ms")
 	WgNormalizationCacheHit            = attribute.Key("wg.engine.normalization_cache_hit")
 	WgValidationCacheHit               = attribute.Key("wg.engine.validation_cache_hit")
+	WgVariablesValidationSkipped       = attribute.Key("wg.engine.variables_validation_skipped")
```
Comment (StarpTech) [inline router/pkg/otel/attributes.go:35]: Do you think this belongs to the `resolver` namespace? See `wg.engine.resolver.wait_time_ms`

File: router/core/graphql_prehandler.go [hunk 8]
```diff
@@ -147,36 +157,18 @@ func (h *PreHandler) Handler(next http.Handler) http.Handler {
 			otel.WgOperationProtocol.String(OperationProtocolHTTP.String()),
 		}
 
-		if h.enableRequestTracing {
```
Comment (StarpTech) [inline router/core/graphql_prehandler.go:150]: `enableRequestTracing` is no longer used in the file.

File: router/core/router.go [hunk 9]
```diff
@@ -333,6 +335,8 @@ func NewRouter(opts ...Option) (*Router, error) {
 		// Required for WunderGraph ART
 		"x-wg-trace",
 		"x-wg-token",
+		"x-wg-skip-loader",
```
Comment (StarpTech) [inline router/core/router.go:338]: Please update the defaults in the json schema

File: router/core/factoryresolver.go [hunk 10]
```diff
@@ -447,6 +450,18 @@ func (l *Loader) Load(engineConfig *nodev1.EngineConfiguration, subgraphs []*nod
 	return &outConfig, nil
 }
 
+func (l *Loader) subgraphName(subgraphs []*nodev1.Subgraph, dataSourceID string) string {
```
Comment (StarpTech) [inline router/core/factoryresolver.go:453]: This can be just a function. We have this logic already in context.go

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**